File: implementations/go/algorithms.go
package aioschema

import (
	"crypto/sha256"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

// ── Regex patterns (§5.3, §5.7, §5.4, §5.2) ─────────────────────────────────

var (
	// HashPattern: (sha256|sha3-256)-[0-9a-f]{64} OR sha384-[0-9a-f]{96}
	HashPattern = regexp.MustCompile(
		`^(sha256|sha3-256)-[0-9a-f]{64}$|^sha384-[0-9a-f]{96}$`,
	)

	// SignaturePattern: ed25519-<128 hex chars>
	SignaturePattern = regexp.MustCompile(`^ed25519-[0-9a-f]{128}$`)

	// AnchorPattern: aios-anchor:<method>:<id>
	AnchorPattern = regexp.MustCompile(`^aios-anchor:[a-z0-9_-]+:[a-zA-Z0-9_-]+$`)

	// TimestampPattern: strict ISO-8601 UTC with Z suffix (§5.2)
	TimestampPattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$`)

	// CreatorIDAttributedPattern: ed25519-fp-<32 hex chars>
	CreatorIDAttributedPattern = regexp.MustCompile(`^ed25519-fp-[0-9a-f]{32}$`)
)

// ── Hash computation ──────────────────────────────────────────────────────────

// ComputeHash hashes data with the named algorithm and returns a prefixed hex string.
// Supported algorithms: "sha256", "sha384", "sha3-256".
// Returns an error for unsupported algorithms.
func ComputeHash(data []byte, algorithm string) (string, error) {
	switch algorithm {
	case "sha256":
		h := sha256.Sum256(data)
		return "sha256-" + hex.EncodeToString(h[:]), nil
	case "sha384":
		h := sha512.Sum384(data)
		return "sha384-" + hex.EncodeToString(h[:]), nil
	case "sha3-256":
		h := sha3_256Sum(data)
		return "sha3-256-" + hex.EncodeToString(h[:]), nil
	default:
		return "", fmt.Errorf("unsupported hash algorithm %q; supported: sha256, sha384, sha3-256", algorithm)
	}
}

// MustComputeHash is like ComputeHash but panics on error (use only in tests).
func MustComputeHash(data []byte, algorithm string) string {
	s, err := ComputeHash(data, algorithm)
	if err != nil {
		panic(err)
	}
	return s
}

// ParseHashPrefix splits "alg-hexdigest" into (algorithm, hexdigest).
// Returns an error if the value does not match the hash pattern.
func ParseHashPrefix(value string) (alg, digest string, err error) {
	if !HashPattern.MatchString(value) {
		return "", "", fmt.Errorf("invalid hash value %q: expected (sha256|sha3-256)-<64hex> or sha384-<96hex>", value)
	}
	// sha3-256 has a hyphen in the algorithm token; handle it first.
	if strings.HasPrefix(value, "sha3-256-") {
		return "sha3-256", value[len("sha3-256-"):], nil
	}
	idx := strings.Index(value, "-")
	return value[:idx], value[idx+1:], nil
}

// safeEqualString performs a timing-safe string comparison (§12.1).
func safeEqualString(a, b string) bool {
	return subtle.ConstantTimeCompare([]byte(a), []byte(b)) == 1
}

// ── Canonical JSON ────────────────────────────────────────────────────────────

// CanonicalJSON returns the deterministic, compact JSON encoding of v with
// all object keys sorted recursively — the canonical form required by §5.6.
// The result is UTF-8 encoded bytes.
func CanonicalJSON(v interface{}) ([]byte, error) {
	sorted := sortValue(v)
	return json.Marshal(sorted)
}

// sortValue recursively sorts map keys to produce canonical JSON.
func sortValue(v interface{}) interface{} {
	switch val := v.(type) {
	case map[string]interface{}:
		// Build a sorted-key wrapper using json.RawMessage to preserve Go's
		// map marshaling sorted by key (Go's encoding/json sorts map keys).
		result := make(map[string]interface{}, len(val))
		for k, elem := range val {
			result[k] = sortValue(elem)
		}
		return result
	case []interface{}:
		result := make([]interface{}, len(val))
		for i, elem := range val {
			result[i] = sortValue(elem)
		}
		return result
	default:
		return v
	}
}

// canonicalCoreFields extracts only the CORE_HASH_FIELDS from a raw core map
// and returns their canonical JSON bytes. Used to compute/verify core_fingerprint.
func canonicalCoreFields(core map[string]interface{}) ([]byte, error) {
	subset := make(map[string]interface{}, len(CoreHashFields))
	for _, field := range CoreHashFields {
		if val, ok := core[field]; ok {
			subset[field] = val
		}
	}
	return CanonicalJSON(subset)
}

// canonicalManifestBytes serializes the entire manifest for manifest_signature
// verification (§5.8). manifest_signature is set to null before serialization
// to avoid circular dependency.
func canonicalManifestBytes(m map[string]interface{}) ([]byte, error) {
	// Deep copy to avoid mutating caller's data
	data, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}
	var copy_ map[string]interface{}
	if err := json.Unmarshal(data, &copy_); err != nil {
		return nil, err
	}
	if core, ok := copy_["core"].(map[string]interface{}); ok {
		core["manifest_signature"] = nil
	}
	return CanonicalJSON(copy_)
}

File: implementations/rust/src/algorithms.rs
//! Hash computation, canonical JSON, and regex patterns for AIOSchema v0.5.5.

use crate::types::{AiosError, CORE_HASH_FIELDS};
use once_cell::sync::Lazy;
use regex::Regex;
use sha2::{Digest, Sha256, Sha384};
use sha3::Sha3_256;
use serde_json::{Map, Value};
use std::collections::BTreeMap;

// ── Compiled regex patterns ───────────────────────────────────────────────────

/// `(sha256|sha3-256)-[0-9a-f]{64}` OR `sha384-[0-9a-f]{96}` (§5.3)
pub static HASH_PATTERN: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^(sha256|sha3-256)-[0-9a-f]{64}$|^sha384-[0-9a-f]{96}$").unwrap()
});

/// `ed25519-<128 hex chars>` (§5.1)
pub static SIGNATURE_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^ed25519-[0-9a-f]{128}$").unwrap());

/// `aios-anchor:<method>:<id>` (§9.1)
pub static ANCHOR_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^aios-anchor:[a-z0-9_-]+:[a-zA-Z0-9_-]+$").unwrap());

/// Strict ISO-8601 UTC with Z suffix (§5.2)
pub static TIMESTAMP_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$").unwrap());

/// `ed25519-fp-<32 hex chars>` attributed creator_id (§5.7)
pub static CREATOR_ID_ATTRIBUTED_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^ed25519-fp-[0-9a-f]{32}$").unwrap());

/// Any well-formed UUID (for anonymous creator_id)
pub static UUID_PATTERN: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$").unwrap()
});

// ── Hash computation ──────────────────────────────────────────────────────────

/// Compute a prefixed hash string `"<alg>-<hex>"` from `data`.
///
/// Supported algorithms: `"sha256"`, `"sha384"`, `"sha3-256"`.
pub fn compute_hash(data: &[u8], algorithm: &str) -> Result<String, AiosError> {
    match algorithm {
        "sha256" => {
            let digest = Sha256::digest(data);
            Ok(format!("sha256-{}", hex::encode(digest)))
        }
        "sha384" => {
            let digest = Sha384::digest(data);
            Ok(format!("sha384-{}", hex::encode(digest)))
        }
        "sha3-256" => {
            let digest = Sha3_256::digest(data);
            Ok(format!("sha3-256-{}", hex::encode(digest)))
        }
        other => Err(AiosError::UnsupportedAlgorithm(other.to_string())),
    }
}

/// Parse `"alg-hexdigest"` into `(algorithm, hex_digest)`.
///
/// Returns `Err` if the value does not match `HASH_PATTERN`.
pub fn parse_hash_prefix(value: &str) -> Result<(&str, &str), AiosError> {
    if !HASH_PATTERN.is_match(value) {
        return Err(AiosError::InvalidHash(format!(
            "invalid hash value {value:?}: expected (sha256|sha3-256)-<64hex> or sha384-<96hex>"
        )));
    }
    // sha3-256 has a hyphen inside the algorithm token — must be checked first.
    if let Some(rest) = value.strip_prefix("sha3-256-") {
        return Ok(("sha3-256", rest));
    }
    // sha256 or sha384
    let dash = value.find('-').unwrap(); // guaranteed by pattern
    Ok((&value[..dash], &value[dash + 1..]))
}

// ── Canonical JSON ────────────────────────────────────────────────────────────

/// Produce deterministic, compact JSON from a `serde_json::Value`.
///
/// All object keys are sorted recursively — this is the canonical form
/// required by §5.6. Uses `BTreeMap` internally so `serde_json` emits
/// sorted keys.
pub fn canonical_json(value: &Value) -> Result<Vec<u8>, AiosError> {
    let sorted = sort_value(value);
    Ok(serde_json::to_vec(&sorted)?)
}

/// Recursively sort all object keys in a JSON value tree.
fn sort_value(value: &Value) -> Value {
    match value {
        Value::Object(map) => {
            let sorted: BTreeMap<String, Value> = map
                .iter()
                .map(|(k, v)| (k.clone(), sort_value(v)))
                .collect();
            // Convert BTreeMap back to serde_json Map (preserves insertion order,
            // but since we iterate a BTreeMap the order is already sorted).
            let mut out = Map::new();
            for (k, v) in sorted {
                out.insert(k, v);
            }
            Value::Object(out)
        }
        Value::Array(arr) => Value::Array(arr.iter().map(sort_value).collect()),
        other => other.clone(),
    }
}

/// Extract exactly `CORE_HASH_FIELDS` from a JSON object and return their
/// canonical bytes. Used to compute and verify `core_fingerprint`.
pub fn canonical_core_fields(core: &Value) -> Result<Vec<u8>, AiosError> {
    let obj = core.as_object().ok_or_else(|| {
        AiosError::Other("core must be a JSON object".to_string())
    })?;

    let mut subset = Map::new();
    for &field in CORE_HASH_FIELDS {
        if let Some(v) = obj.get(field) {
            subset.insert(field.to_string(), v.clone());
        }
    }
    canonical_json(&Value::Object(subset))
}

/// Serialize a full manifest for `manifest_signature` verification (§5.8).
///
/// `manifest_signature` is set to `null` before serialization to avoid a
/// circular dependency (bootstrap exclusion).
pub fn canonical_manifest_bytes(manifest: &Value) -> Result<Vec<u8>, AiosError> {
    let mut m = manifest.clone();
    if let Some(core) = m.get_mut("core").and_then(|c| c.as_object_mut()) {
        core.insert("manifest_signature".to_string(), Value::Null);
    }
    canonical_json(&m)
}

// ── Timing-safe comparison ────────────────────────────────────────────────────

/// Timing-safe string equality (§12.1).
pub fn safe_equal(a: &str, b: &str) -> bool {
    // Constant-time comparison via XOR fold.
    let ab = a.as_bytes();
    let bb = b.as_bytes();
    if ab.len() != bb.len() {
        return false;
    }
    let mut diff: u8 = 0;
    for (x, y) in ab.iter().zip(bb.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}
